code/modules/clothing/head/perceptomatrix.dm

```diff
@@ -0,0 +1,202 @@
+		granting the user effective immunity to both psychic threats, and anything that would affect their perception - be it ear, eye, or even brain damage. \
+		It can also violently discharge said energy, inducing hallucinations in others."
+	icon_state = "perceptomatrix_helmet_inactive"
+	worn_icon_state = "perceptomatrix_helmet_inactive"
+	base_icon_state = "perceptomatrix_helmet"
+	force = 10
+	dog_fashion = null
+	cold_protection = HEAD
+	min_cold_protection_temperature = HELMET_MIN_TEMP_PROTECT
+	heat_protection = HEAD
+	max_heat_protection_temperature = HELMET_MAX_TEMP_PROTECT
+	strip_delay = 8 SECONDS
+	clothing_flags = SNUG_FIT|STACKABLE_HELMET_EXEMPT|STOPSPRESSUREDAMAGE|BLOCK_GAS_SMOKE_EFFECT | CASTING_CLOTHES // we love casting spells
+	clothing_traits = list(
+		/* eye/ear protection */
+		TRAIT_NOFLASH,
+		TRAIT_TRUE_NIGHT_VISION,
+		TRAIT_SIGHT_BYPASS,
+		TRAIT_EXPANDED_FOV,
+		TRAIT_GOOD_HEARING,
+		/* mental protection */
+		TRAIT_PERCEPTUAL_TRAUMA_BYPASS, //wip
+		TRAIT_MADNESS_IMMUNE,
+		/* psychic protection */
+		TRAIT_NO_MINDSWAP,
+		TRAIT_UNCONVERTABLE,
+	)
+	flags_cover = HEADCOVERSEYES|EARS_COVERED
+	flags_inv = HIDEHAIR|HIDEFACE
+	flash_protect = FLASH_PROTECTION_WELDER_SENSITIVE
+	resistance_flags = FIRE_PROOF | ACID_PROOF
+	equip_sound = 'sound/items/handling/helmet/helmet_equip1.ogg'
+	pickup_sound = 'sound/items/handling/helmet/helmet_pickup1.ogg'
+	drop_sound = 'sound/items/handling/helmet/helmet_drop1.ogg'
+	armor_type = /datum/armor/head_helmet_matrix
+	actions_types = list(/datum/action/cooldown/spell/pointed/percept_hallucination)
+
+	/// If we have a core or not
+	var/core_installed = FALSE
+	var/list/active_components
+
+// weaker overall but better against energy
+/datum/armor/head_helmet_matrix
+	melee = 15
+	bullet = 15
+	laser = 45
+	energy = 60
+	bomb = 15
+	fire = 50
+	acid = 50
+	wound = 10
+
+/obj/item/clothing/head/helmet/perceptomatrix/Initialize(mapload)
+	. = ..()
+
+	update_icon_state()
+	update_anomaly_state()
+
+	// some-fucking-how, RemoveElement fails to find the element in below proc if i try to remove it. so whatever.
+	AddElement(/datum/element/wearable_client_colour, /datum/client_colour/perceptomatrix, ITEM_SLOT_HEAD, forced = TRUE)
+
+/obj/item/clothing/head/helmet/perceptomatrix/proc/update_anomaly_state()
+
+	// If the core isn't installed, or it's temporarily deactivated, disable special functions.
+	if(!core_installed)
+		clothing_flags = initial(clothing_flags) & ~CASTING_CLOTHES
+		detach_clothing_traits(clothing_traits)
+		QDEL_LIST(active_components)
+		return
+
+	clothing_flags = initial(clothing_flags)
+	attach_clothing_traits(initial(clothing_traits))
+
+	LAZYADD(active_components, AddComponent(/datum/component/wearertargeting/earprotection, list(ITEM_SLOT_HEAD)))
+	LAZYADD(active_components, AddComponent(
+		/datum/component/anti_magic, \
+		antimagic_flags = MAGIC_RESISTANCE_MIND, \
+		inventory_flags = ITEM_SLOT_HEAD, \
+	))
```

Comment:
`active_components` being a lazylist seems overcomplex

Reply:
gotta instance the list somehow